art/defences/trainer/ibp_certified_trainer_pytorch.py

```diff
@@ -0,0 +1,462 @@
+                if self._classifier._use_amp:  # pragma: no cover # pylint: disable=W0212
+                    from apex import amp  # pylint: disable=E0611
+
+                    with amp.scale_loss(loss, self._classifier._optimizer) as scaled_loss:  # pylint: disable=W0212
+                        scaled_loss.backward()
+
+                else:
+                    loss.backward()
+
+                self._classifier._optimizer.step()  # pylint: disable=W0212
+                self._classifier.re_convert()
+
+                if verbose:
+                    pbar.set_description(
+                        f"Bound {bound:.2f}: "  # pylint: disable=W0212
+                        f"Loss {torch.mean(torch.stack(epoch_non_cert_loss)):.2f} "
+                        f"Cert Loss {torch.mean(torch.stack(cert_loss)):.2f} "
+                        f"Acc {np.mean(non_cert_acc):.2f} Cert Acc {np.mean(cert_acc):.2f} "
+                        f"l_weight {loss_weighting_k:.2f} lr {self._classifier._optimizer.param_groups[0]['lr']}"
+                    )
+
+            if scheduler is not None:
+                scheduler.step()
+
+    def predict(self, x: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Perform prediction using the adversarially trained classifier.
+
+        :param x: Input samples.
+        :param kwargs: Other parameters to be passed on to the `predict` function of the classifier.
+        :return: Predictions for test set.
+        """
+        if self._classifier._model._model.forward_mode != "concrete":  # pylint: disable=W0212
+            raise ValueError(
+                "For normal predictions, the model must be running in concrete mode. If an abstract "
+                "prediction is wanted then use predict_interval instead"
+            )
+
+        return self._classifier.predict(x, **kwargs)
+
+    def predict_intervals(
+        self,
+        x: np.ndarray,
+        is_interval: bool = False,
+        bounds: Optional[Union[float, List[float], np.ndarray]] = None,
+        limits: Optional[Union[List[float], np.ndarray]] = None,
+        batch_size: int = 128,
+        **kwargs,
+    ) -> np.ndarray:
+        """
+        Perform prediction using the adversarially trained classifier using zonotopes
+
+        :param x: The datapoint, either:
+
+                1. In the interval format of x[batch_size, 2, feature_1, feature_2, ...]
+                   where axis=1 corresponds to the [lower, upper] bounds.
+
+                2. Or in regular concrete form, in which case the bounds/limits need to be supplied.
+        :param is_interval: if the datapoint is already in the correct interval format.
+        :param bounds: The perturbation range.
+        :param limits: The clipping to apply to the interval data.
+        :param batch_size: batch size to use when looping through the data
+        """
+
+        if self._classifier._model._model.forward_mode != "abstract":  # pylint: disable=W0212
```

Comment:
Is this required, it should be possible to do
```suggestion
        if self.classifier.model.forward_mode != "abstract":
```

Reply:
I agree! Change implemented.